RETURNING clause for postgresql

Zeliboba5 opened the pull request:
This PR implements postgres `RETURNING` clause.
This PR implements it directly in QueryBuilder, which could be quite misleading, for other sql users, so if you have any propositions regarding how this code should be arranged I can do some refactoring

twheys commented:
If you'd like to add a vendor specific function you can override the querybuilder class and instantiate it in the PostgreSQLQuery class' _builder function. See my last PR for an example: https://github.com/kayak/pypika/pull/109

twheys commented on pypika/queries.py:
```diff
@@ -570,6 +584,49 @@ def _select_other(self, function):
+
+    def _validate_returning_term(self, term):
+        if not any([self._insert_table, self._update_table, self._delete_from]):
+            QueryException('Returning can\'t be used in this query')
```
It looks like you're missing the `raise` keyword to raise this exception

twheys commented on pypika/queries.py:
```diff
@@ -570,6 +584,49 @@ def _select_other(self, function):
+                term.table not in {self._insert_table, self._update_table}
+                and term not in self._from
+        ):
+            QueryException('You can\'t return from other tables')
```
It looks like you're missing the `raise` keyword to raise this exception

twheys commented on pypika/queries.py:
```diff
@@ -442,6 +444,18 @@ def having(self, criterion):
             self._havings = criterion
 
+    @builder
+    def returning(self, *terms):
```
This code looks very similar to the `select` function. Would it be possible to consolidate the similarities between the two?

twheys commented on pypika/tests/test_inserts.py:
```diff
@@ -88,6 +88,35 @@ def test_insert_null(self):
         self.assertEqual('INSERT INTO "abc" VALUES (NULL)', str(query))
 
 
+class InsertIntoReturningTests(unittest.TestCase):
```
These test cases look pretty thorough and I'm not sure what's missed but coveralls is reporting 0.5% drop in test case coverage. Can you please check and make sure all of your conditions are covered by test cases?

twheys commented:
Sorry, not sure how I closed this, didn't mean to. Re-opened.

Zeliboba5 commented:
Moved everything to postgres QueryBuilder, added some more tests and covered some more clauses, now it all looks neat.
Regarding consolidation of `returning` and `select` I think it is wrong way to go, as it still different functions, which are not fully compatible (`RETURNING` does not support aggregate functions, so I covered it in my new commits)

twheys commented:
Looks good to go, just need to fix the merge conflicts. I'll merge once that's done, let me know if you need help with it.

twheys commented:
Merged and published 0.12.0
